MetaU Capstone Project/API/SpotifyManager.m

```diff
@@ -247,6 +260,144 @@ - (void)searchTrack:(NSString *)query type:(NSString *)type result:(void (^)(NSD
         NSData *dataUTF8 = [strISOLatin dataUsingEncoding:NSUTF8StringEncoding];
         dataRecieved = [NSJSONSerialization JSONObjectWithData:dataUTF8 options:0 error:&error];
         
+        if (dataRecieved != nil) {
+            parsingFinished([dataRecieved copy]);
+        } else {
+            NSLog(@"Error: %@", error.localizedDescription);
+            parsingFinished([[NSDictionary alloc] init]);
+        }
+    }] resume];
+}
+
+- (void)addQueueToSpotify: (NSString *)trackURI {
+    [[self.appRemote playerAPI] enqueueTrackUri:trackURI callback:^(id  _Nullable result, NSError * _Nullable error) {
+            if (error != nil) {
+                NSLog(@"Error: %@", error.localizedDescription);
+            } else {
+                NSLog(@"%@ added to Spotify queue", trackURI);
+            }
+    }];
+}
+
+- (NSDictionary *)currentTrackInfo: (id<SPTAppRemoteTrack>)track {
+    NSDictionary *trackDetails = @{
+            @"name" : track.name,
+            @"URI" : track.URI,
+            @"contextURI" : track.album.URI,
+            @"artist" : track.artist.name,
+            @"album" : track.album.name,
+            @"images" : @[]
+    };
+
+    return (NSDictionary *)trackDetails;
+}
+
+- (NSDictionary *)getCurrentTrack {
+    return self.currentTrack;
+}
+
+- (id<SPTAppRemoteTrack>)getCurrentTrackAsContentItem {
+    return self.currentTrackContentItem;
+}
+
+- (void)recentlyPlayedTrack {
+    NSString *targetURL = @"https://api.spotify.com/v1/me/player/recently-played?limit=1";
+    
+    [self retriveDataFrom:targetURL result:^(NSDictionary * _Nonnull dataRecieved) {
+        NSString *track = [[dataRecieved valueForKey:@"items"] valueForKey:@"track"][0];
```

Comment:
I haven't checked how the `retriveDataFrom:` function is implemented, but I don't think `dataReceived` is always going to be nonnull.
This assumption comes from the fact that you're loading JSON from some external API endpoint.
That is usually never a *confirmed-to-always-work* operation. In the cases where it fails, the JSON you expect is likely going to be nil.

Calling `valueForKey:` on a null dictionary can cause a crash (It happened to me recently)

Anyway, I would suggest that you:
- remove the nonnull specifier
- first check that `dataReceived` really is nonnull
- Then use as many if statements as you need to check if you can get `track` without crashing